app/packages/core/src/components/Grid/useExpandSample.ts

```diff
@@ -79,6 +66,27 @@ export default <T extends fos.Lookers>(store: fos.LookerStore<T>) => {
         itemIndexMap: ItemIndexMap
       ): Promise<void> => {
         const clickedIndex = itemIndexMap[sampleId];
+        // Cache the page parameters so that we don't need `snapshot` (which
+        // is released once the useRecoilCallback ends) to async load samples.
+        const pageParams = await snapshot.getPromise(
+          pageParameters(/* modal= */ true)
+        );
+
+        // Get the full sample if the existing sample has thumbnails only.
+        // Then set it in the modal if the user hasn't navigated away.
+        const getAndSetFullSampleIfNeeded = (
+          existingSample: fos.SampleData
+        ) => {
+          if (existingSample.thumbnailsOnly) {
+            const sampleId = existingSample.sample._id;
+            getFullSample(store, pageParams, sampleId).then((sample) => {
+              // The ID of the sample in the modal must match the queried sample
+              if (currentSampleId.current === sampleId) {
+                setSample(sample, { index: clickedIndex, getIndex });
```

Comment:
What's going to happen here if I select another sample while this sample is being loaded?

Reply:
* If the sample finishes loading before the new click is handled:
  * `setSample` will be called here, and the modal may briefly show the old sample before moving to the new one
* If the new click is handled before the sample is loaded:
  * `currentSampleId.current` will be updated to the new sampleId
  * `setSample` won't be called on the old sample since `currentSampleId` has changed

In the updated video I spammed the "right" button and it loads 3 samples at a time, but stays on the last sample.